feat(patterns): prepare main codebase for blog

cerebrl opened the pull request:
## Summary

Deep refactor of React Native codebase for preparation of blog articles and education:

- Improved UX
- Improved Error handling
- Better visual alignment with React JS app
- Additional comments along with debugger statements
- Reduction of abstractions and indirections for easier understanding

## Code Review

It is recommended that you just download this branch onto a local computer and review the React Native code while running it. This refactor's too big to review via PR. Sorry!

## Screen Recording of UX

![React Native UX part1](https://user-images.githubusercontent.com/3070458/142562765-8ddaf7a3-f599-4375-ba6b-564bf6705825.gif)

![React Native UX part2](https://user-images.githubusercontent.com/3070458/142562775-4999e763-0f1d-4035-ad6b-e1f3f6acad3d.gif)

ryanbas21 reviewed (approved):
Built and run, read through code locally. Looks good to me

cerebrl commented:
Split Swift bridge code into two files: one file for the bridge class (`FRAuthSampleBridge.swift`) and the other file with the types/struct/enum definitions (`FRAuthSampleTypes.swift`). I also added verbose code comments to help aid the reader in understanding the logic/intent of the Swift code as well as breaking some of the very long lines to help avoid horizontal scrolling.

george-bafaloukas-forgerock reviewed (changes requested):
We might need to do 1 extra change for the Podfile. Will confirm and add a new comment

cerebrl commented on reactnative-todo/ios/reactnativetodo/FRAuthSampleBridge.swift:
```diff
@@ -147,61 +178,84 @@ public class FRAuthSampleBridge: NSObject {
+            resolve(["type": "LoginSuccess", "tokens": jsonAccessToken])
           } else {
-            resolve(["type": "LoginSuccess", "sessionToken": ""])
+            resolve(["type": "LoginSuccess", "tokens": ""])
```
I changed the UI to expect `tokens`. The use of `sessionToken` is a bit misleading since it contains all the tokens.

george-bafaloukas-forgerock replied:
Makes sense

cerebrl commented:
@george-bafaloukas-forgerock & @vahancouver: PR has been updated with improved comments and more Swifty style :) Please review and test!

vahancouver commented on reactnative-todo/ios/reactnativetodo/FRAuthSampleStructs.swift:
```diff
@@ -0,0 +1,250 @@
+  case Bool
+  case TypeMismatch
+  case NotSet
+}
```
In swift enum cases naming convention is lowerCamelCase (I think before Swift 3 it was UpperCamelCase)

george-bafaloukas-forgerock replied:
Good point

vahancouver commented on reactnative-todo/ios/reactnativetodo/FRAuthSampleStructs.swift:
```diff
@@ -0,0 +1,250 @@
+public struct RawCallback: Codable {
+  var type: String?
+  var input: [RawInput]?
+  var _id: Int?
```
Usually in swift we never start variable names with underscore.

george-bafaloukas-forgerock replied:
I agree, but this is to parse the response as it is coming from AM. I could have the coding keys but this is minimal impact for the rest of the code. Also, this Codable is only used for a very specific reason for parsing responses in the bridge

vahancouver commented on reactnative-todo/ios/reactnativetodo/FRAuthSampleStructs.swift:
```diff
@@ -0,0 +1,250 @@
+
+    if let thisCallback = callback as? SingleValueCallback {
+      self.prompt = thisCallback.prompt
+      self.inputNames = [thisCallback.inputName!]
```
Is it always guaranteed for the callback inputName have a value? I'm a little afraid of using forced unwrapping. Maybe we could have provided a default value instead:   `self.inputNames = [thisCallback.inputName ?? "default_value"]

george-bafaloukas-forgerock replied:
Good catch this has been missed, it shouldn't be like that and will be changed

vahancouver reviewed (commented):
Added some comments related to Swift

spetrov reviewed (approved):
I followed the tutorial and ran the todo app on iOS. Everything went smoothly, except for a minor problem when I tried to start the API server using `npm run server`. It turned out that I needed to add an additional property in the package.json file `"type": "module"`.
